feat(discover) - Migrate v1 queries into v2

wmak opened the pull request:
- Take the existing discover v1 queries and create equivalent
  discover v2 queries.
- The only things we should be losing is:
	- There's no equivalent to `limit` in v2
	- From a conversation with mark contexts.key is too expensive
	  and should be removed

wmak committed:
feat(discover) - Migrate v1 queries into v2

- Take the 1,283 existing discover v1 queries and create equivalent
  discover v2 queries.
- The only things we should be losing is:
	- There's no equivalent to `limit` in v2
	- From a conversation with mark contexts.key is too expensive
	  and should be removed

wmak commented on src/sentry/migrations/0028_discover_query_upgrade.py:
```diff
@@ -0,0 +1,150 @@
+        value = u'"{}"'.format(value)
+    return value
+
+def convert(saved_query, name_extra=" (migrated from legacy discover)"):
```
I added an extra bit to the end of each name to make it clearer where these new queries are coming from: 
![image](https://user-images.githubusercontent.com/4205004/73395654-f119d980-42ad-11ea-9526-49cb0b85f69e.png)

markstory replied:
Nice touch :heart:

markstory reviewed (commented):
This looks great. From your testing how many existing queries were not able to be converted?

wmak commented:
@markstory I actually didn't see any queries that weren't able to be converted -- But that's from me visually inspecting the differences between old/new queries.

evanh commented on src/sentry/migrations/0028_discover_query_upgrade.py:
```diff
@@ -0,0 +1,150 @@
+    if fieldname == "count":
+        fieldname = u"count()"
+    elif unique:
+        fieldname = u"count_unique({})".format(fieldname) if unique else field
```
You don't need the `if unique` here.

wmak replied:
🙃 the else was on an old variable name too -- good catch

evanh commented on src/sentry/migrations/0028_discover_query_upgrade.py:
```diff
@@ -0,0 +1,150 @@
+
+    if "range" in saved_query.query:
+        updated_query["range"] = saved_query.query["range"]
+    elif "start" in saved_query.query and "end" in saved_query.query:
```
If there is a start is there always guaranteed to be an end? You can't have a start and a range?

wmak replied:
Based on the data, the 68 rows with starts always have an `end` :+1:
There's 5 with `start` and a `range` but range is an empty string in those cases

dashed replied:
I expect start, end, and range to be converted in the same precedence order as this function https://github.com/getsentry/sentry/blob/d199c9a2dda5bdf85627291d216d5caecd960222/src/sentry/api/utils.py#L33-L84

wmak replied:
TBH I could probably just forcibly use that function, but there doesn't exist any queries that require using it. And I think it makes more sense if we use range over start/end since start/ends stop working once they go beyond retention.

dashed replied:
Ah. I didn't mean to suggest that this function be used. 😅 

I think the implementation you had was fine. 👍

bowencai8 commented:
Thanks William! What is contexts.key used for? and do you know how many saved queries had `limit`?

wmak commented:
@bowencai8 it contains string lists of all the keys/values for contexts and tags
So I think limit was required in v1 as there's only 5 without a limit, but most of them have a limit of 1000 which is the default, and there are 61 that have a different limit.

bowencai8 commented:
@wmak Got it, thanks!

dashed commented on src/sentry/migrations/0029_discover_query_upgrade.py:
```diff
@@ -0,0 +1,150 @@
+    queryset = DiscoverSavedQuery.objects.filter(version=1)
+
+    for query in RangeQuerySetWrapperWithProgressBar(queryset):
+        new_query = convert(query)
```
`new_query` variable seems to be unused here. 👀 Does this do anything?

You don't need to fix this; I was just curious.

wmak replied:
Ah woops, that's a relic from me running tests on the migration... I'm... not going to fix this cause I don't want to wait for Travis to pass again 🙃

mattrobenolt commented on src/sentry/migrations/0029_discover_query_upgrade.py:
```diff
@@ -0,0 +1,150 @@
+from __future__ import unicode_literals, print_function
+from sentry.utils.query import RangeQuerySetWrapperWithProgressBar
+from sentry.discover.models import DiscoverSavedQuery, DiscoverSavedQueryProject
+from sentry.api.event_search import SearchVisitor
```
Neither of these can safely be imported here. They're very easily subject to breaking in the future and breaking this migration.

Ask @wedamija how to reference the models safely since I'm not sure how we do it in Django migrations, but we need to access them through the frozen state.

And for SearchVisitor, it looks like you're just referencing a couple constants. You should vendor those constants into this migration.

wmak committed:
ref(imports) - Removing imports of models + SearchVisitor

- Moving constants that were used from SearchVisitor directly into this
  file
- Using apps.get_model to get the Discover models instead of directly
  importing them into the Migration

wedamija commented:
How long does this migration take to run? You may want to run it with `atomic = False` so that it doesn't run in a single transaction.

wmak commented on src/sentry/migrations/0029_discover_query_upgrade.py:
```diff
@@ -0,0 +1,201 @@
+    #   they can be monitored. Since data migrations will now hold a transaction open
+    #   this is even more important.
+    # - Adding columns to highly active tables, even ones that are NULL.
+    is_dangerous = False
```
Locally this migration runs in ~30s
@wedamija do you mean like this?
```suggestion
    is_dangerous = False
    atomic = False
```

wedamija replied:
@wmak Yep! Yeah, I think it's a good idea to mark it as non-atomic.

wmak committed:
ref: Add atomic=False

wedamija reviewed (commented):
I didn't check the logic of the conversion itself, but it should be good to go assuming that is correct.
